crates/bevy_ecs/src/schedule/runner/mod.rs

```diff
@@ -0,0 +1,190 @@
+use fixedbitset::FixedBitSet;
+
+/// Types that can run [`System`] instances on the data stored in a [`World`].
+pub trait SystemRunner: Downcast + Send + Sync {
+    fn run(&mut self, world: &mut World);
+}
+
+impl_downcast!(SystemRunner);
+
+/// Internal resource used to signal the runner to apply pending commands to the [`World`].
+pub(super) struct RunnerApplyBuffers(pub bool);
+
+impl Default for RunnerApplyBuffers {
+    fn default() -> Self {
+        Self(false)
+    }
+}
+
+/// Applies pending command queues to the [`World`].
+pub fn apply_buffers(world: &mut World) {
+    // Best place to do this check.
+    world.check_change_ticks();
+
+    let mut should_apply_buffers = world
+        .get_resource_mut::<RunnerApplyBuffers>()
+        .expect("RunnerApplyBuffers resource does not exist.");
+
+    // The executor resets this to false, so if it's true, something is weird.
+    assert!(
+        !should_apply_buffers.0,
+        "pending commands should have been applied"
+    );
+    should_apply_buffers.0 = true;
+}
+
+pub(crate) struct Runner {
+    /// A list of systems, topsorted according to system dependency graph.
+    pub(crate) systems: Vec<BoxedSystem>,
+    /// A list of system run criteria, topsorted according to system dependency graph.
+    pub(crate) system_conditions: Vec<Vec<BoxedRunCondition>>,
+    /// A list of set run criteria, topsorted according to set hierarchy graph.
+    pub(crate) set_conditions: Vec<Vec<BoxedRunCondition>>,
+    /// A map relating each system index to its number of dependencies and list of dependents.
+    pub(crate) system_deps: HashMap<usize, (usize, Vec<usize>)>,
+    /// A map relating each system index to a bitset of its ancestor sets.
+    pub(crate) system_sets: HashMap<usize, FixedBitSet>,
+    /// A map relating each set index to a bitset of its descendant systems.
+    pub(crate) set_systems: HashMap<usize, FixedBitSet>,
+    /// A list of systems (their ids), topsorted according to dependency graph.
+    pub(crate) systems_topsort: Vec<RegistryId>,
+    /// A list of sets (their ids), topsorted according to hierarchy graph.
+    pub(crate) sets_topsort: Vec<RegistryId>,
+}
+
+impl Default for Runner {
+    fn default() -> Self {
+        Self {
+            systems_topsort: Vec::new(),
+            systems: Vec::new(),
+            system_conditions: Vec::new(),
+            system_deps: HashMap::new(),
+            system_sets: HashMap::new(),
+            sets_topsort: Vec::new(),
+            set_conditions: Vec::new(),
+            set_systems: HashMap::new(),
+        }
+    }
+}
+
+impl Runner {
+    pub(crate) fn simple(self) -> SimpleRunner {
+        SimpleRunner::with(self)
+    }
+
+    pub(crate) fn single_threaded(self) -> SingleThreadedRunner {
+        SingleThreadedRunner::with(self)
+    }
+}
+
+/// Runs the system or system set given by `schedule` on the world.
+pub fn run_systems(schedule: impl SystemLabel, world: &mut World) {
```

Comment:
I'd like to see more expects in this function; this doesn't look like fun to debug when tweaking.